File: competitive-programming/leet-code/313/nthSuperUglyNumber.js
var nthSuperUglyNumber = function (n, primes) {
    const ugly = new Array(n).fill(1);
    const idx = new Array(primes.length).fill(0);
    const val = primes.slice();

    for (let i = 1; i < n; i++) {
        const minVal = Math.min(...val);
        ugly[i] = minVal;

        for (let j = 0; j < primes.length; j++) {
            if (val[j] === minVal) {
                idx[j]++;
                val[j] = ugly[idx[j]] * primes[j];
            }
        }
    }

    return ugly[n - 1];
};

File: competitive-programming/leet-code/313/nth_super_ugly_number.py
from typing import List

def nth_super_ugly_number(n: int, primes: List[int]) -> int:
    ugly = [1] * n
    idx = [0] * len(primes)
    val = primes.copy()

    for i in range(1, n):
        min_val = min(val)
        ugly[i] = min_val

        for j in range(len(primes)):
            if val[j] == min_val:
                idx[j] += 1
                val[j] = ugly[idx[j]] * primes[j]

    return ugly[n - 1]
